lib/pos_mempool.go

```diff
@@ -277,6 +277,88 @@ func (mp *PosMempool) IsRunning() bool {
 	return mp.status == PosMempoolStatusRunning
 }
 
+// OnBlockConnected is an event handler provided by the PoS mempool to handle the blockchain
+// event where a block is connected to the tip of the blockchain. The mempool updates its
+// internal state based on the new block that has been connected.
+//
+// Whenever a block is connected, this event handler removes the block's transactions from
+// the mempool and updates the internal fee estimation to include new block.
+func (mp *PosMempool) OnBlockConnected(blockEvent *BlockEvent) {
+	mp.Lock()
+	defer mp.Unlock()
+
+	if blockEvent == nil || blockEvent.Block == nil || len(blockEvent.Block.Txns) == 0 {
+		return
+	}
+
+	if !mp.IsRunning() {
+		return
+	}
+
+	// Remove all transactions in the block from the mempool.
+	for _, txn := range blockEvent.Block.Txns {
+		txnHash := txn.Hash()
+
+		// This should never happen. We perform a nil check on the txn hash to avoid a panic.
+		if txnHash == nil {
+			continue
+		}
+
+		// Get the transaction from the register. If the txn doesn't exist in the register,
+		// then there's nothing left to do.
+		txn := mp.txnRegister.GetTransaction(txnHash)
+		if txn == nil {
+			continue
+		}
+
+		mp.removeTransactionNoLock(txn, true)
+	}
+}
+
+// OnBlockDisconnected is an event handler provided by the PoS mempool to handle the blockchain
+// event where a block is disconnected from the tip of the blockchain. The mempool updates its
+// internal state based on the new block that has been connected.
+//
+// Whenever a block is disconnected, this event handler adds the block's transactions back to
+// the mempool and updates the internal fee estimation to exclude the disconnected block.
+func (mp *PosMempool) OnBlockDisconnected(blockEvent *BlockEvent) {
+	mp.Lock()
+	defer mp.Unlock()
+
+	if blockEvent == nil || blockEvent.Block == nil || len(blockEvent.Block.Txns) == 0 {
+		return
+	}
+
+	if !mp.IsRunning() {
+		return
+	}
+
+	// Remove all transactions in the block from the mempool.
+	for _, txn := range blockEvent.Block.Txns {
+		txnHash := txn.Hash()
+
+		// This should never happen. We perform a nil check on the txn hash to avoid a panic.
+		if txnHash == nil {
+			continue
+		}
+
+		// Construct the MempoolTx from the MsgDeSoTxn.
+		mempoolTx, err := NewMempoolTx(txn, blockEvent.Block.Header.TstampNanoSecs/1000, mp.latestBlockHeight)
+		if err != nil {
+			continue
+		}
+
+		// Add the transaction to the mempool and then prune if needed.
+		if err := mp.addTransactionNoLock(mempoolTx, true); err != nil {
+			glog.Errorf("PosMempool.AddTransaction: Problem adding transaction to mempool: %v", err)
+		}
+	}
+
+	if err := mp.pruneNoLock(); err != nil {
+		glog.Errorf("PosMempool.AddTransaction: Problem pruning mempool: %v", err)
+	}
```

Comment:
We'll want to call `RemoveBlock` on `mp.feeEstimator` somewhere in here as well.